test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,1567 @@
+	"iSCSI [Feature:Volumes]":    initISCSIDriver,
+	"Ceph-RBD [Feature:Volumes]": initRbdDriver,
+	"CephFS [Feature:Volumes]":   initCephFSDriver,
+	"hostpath":                   initHostpathDriver,
+	"hostpathSymlink":            initHostpathSymlinkDriver,
+	"emptydir":                   initEmptydirDriver,
+	"cinder":                     initCinderDriver,
+	"gce":                        initGceDriver,
+	"vSphere":                    initVSphereDriver,
+	"azure":                      initAzureDriver,
+	"aws":                        initAwsDriver,
+}
+
+var testCases = map[string]func() testCase{
+	"volumes":                          initVolumesTestCase,
+	"volumeIO":                         initVolumeIOTestCase,
+	"volumeMode [Feature:BlockVolume]": initVolumeModeTestCase,
+	"subPath":                          initSubPathTestCase,
+}
+
+var _ = utils.SIGDescribe("Non-CSI", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(cs)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
+			ServerNodeName: node.Name,
+		}
+	})
+
+	for testCaseName, testCaseInit := range testCases {
+		curTestCaseName := testCaseName
+		curTestCase := testCaseInit()
+		testPatterns := curTestCase.getTestPatterns()
+		curTestFunc := curTestCase.getTestFunc()
+
+		for testPatternName, testPattern := range testPatterns {
+			curTestPatternName := testPatternName
+			curTestPattern := testPattern
+
+			for driverName, initDriver := range testDrivers {
+				curDriverName := driverName
+				curInitDriver := initDriver
+				Context(fmt.Sprintf("Driver: %s, Test: %s, TestPattern: %s",
+					curDriverName, curTestCaseName, curTestPatternName), func() {
+					var (
+						driver       testDriver
+						testCase     testCase
+						testResource testResource
+						testInput    testInput
+					)
+
+					BeforeEach(func() {
+						driver = curInitDriver(f, config, node)
+						// Skip tests before driver creates test resource to avoid unnecessary initialization
+						skipTestForDriver(curTestPattern, driver)
+
+						driver.createDriver()
+						testCase = curTestCase
+						testResource = testCase.initTestResource()
+						testResource.setup(driver, curTestPattern)
+						testInput = testCase.createTestInput(curTestPattern, testResource)
+					})
+
+					AfterEach(func() {
+						testResource.cleanup()
+						driver.cleanupDriver()
+					})
+
+					curTestFunc(&testInput)
+				})
+			}
+		}
+	}
+})
+
+type testDriver interface {
+	createDriver()
+	cleanupDriver()
+	getDriverInfo() driverInfo
+}
+
+type inlineVolumeTestDriver interface {
+	testDriver
+	getVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
```

Comment:
I think for inline volume and static pv interfaces, there needs to be one more call to createVolume, and deleteVolume, instead of being done as part of create/delete driver.  Because for the dynamic provisioning case, you shouldn't need to manually create the volume

Reply:
Separated them as suggested.

For drivers tested with testServer pod, like NFS, iSCSI, and Ceph, volumes are predefined in container images. So, they can't separate ```createDriver``` logic and ```createVolume``` logic.
In addition, dynamic provision pod for NFS seems to be developed separately from that for staticPV/inlineVolume, therefore a different pod from a different container image needs to be launched.  
(This means that two pods, for staticPV/inlineVolume and dynamicPV, will be launched for NFS in dynamic provision test, if we put testServer creation logic in ```createDriver```.)

Therefore, I put testServer creation logic for these drivers in ```createVolume```, not in ```createDriver```.
(I think testServer creation for these drivers include volume creation, so it would be acceptable.)

Note that most of in-tree drivers do nothing in ```createDriver```, however all csi-drivers will need it to create their driver pods, so I think that it makes sence to keep them separate for them.